File: python3/Backtracking/find_all_subsets.py
from typing import List


def find_all_subsets(nums: List[int]) -> List[List[int]]:
    res = []
    backtrack(0, [], nums, res)
    return res

def backtrack(i: int, curr_subset: List[int], nums: List[int], res: List[List[int]]) -> None:
    # Base case: if all elements have been considered, add the
    # current subset to the output.
    if i == len(nums):
        res.append(curr_subset[:])
        return
    # Include the current element and recursively explore all paths
    # that branch from this subset.
    curr_subset.append(nums[i])
    backtrack(i + 1, curr_subset, nums, res)
    # Exclude the current element and recursively explore all paths
    # that branch from this subset.
    curr_subset.pop()
    backtrack(i + 1, curr_subset, nums, res)

File: java/Backtracking/FindAllSubsets.java
import java.util.ArrayList;
import java.util.List;

public class FindAllSubsets {
    public List<List<Integer>> findAllSubsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        backtrack(0, new ArrayList<>(), nums, res);
        return res;
    }

    private void backtrack(int i, List<Integer> currSubset, int[] nums, List<List<Integer>> res) {
        // Base case: if all elements have been considered, add the
        // current subset to the output.
        if (i == nums.length) {
            res.add(new ArrayList<>(currSubset));
            return;
        }
        // Include the current element and recursively explore all paths
        // that branch from this subset.
        currSubset.add(nums[i]);
        backtrack(i + 1, currSubset, nums, res);
        // Exclude the current element and recursively explore all paths
        // that branch from this subset.
        currSubset.remove(currSubset.size() - 1);
        backtrack(i + 1, currSubset, nums, res);
    }
}
